make shared_helpers exe function work for both cygwin and non-cygwin hosts

jeremyd2019 opened the pull request:
On Cygwin, it needs to not append .exe, because /proc/self/exe (and therefore `std::env::current_exe`) does not include the .exe extension, breaking bootstrap's rustc wrapper.  On hosts other than Cygwin, it *does* need to append .exe because the file really does have a .exe extension, and non-Cygwin hosts won't be doing the same filename rewriting that Cygwin does when looking for a file X but finding only X.exe in its place.

Arising from discussion in https://github.com/rust-lang/rust/pull/140154#pullrequestreview-2855782812
@mati865 @Berrysoft

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15175321061/job/42674450723) [(plain)](https://github.com/rust-lang/rust/commit/69b8ffe03c6ca85c6f8c2f950ac95d3a7ca5489b/checks/42674450723/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.841 Building wheels for collected packages: reuse
#12 2.842   Building wheel for reuse (pyproject.toml): started
#12 3.062   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.063   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#12 3.063   Stored in directory: /tmp/pip-ephem-wheel-cache-ad3zwcfr/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.066 Successfully built reuse
#12 3.066 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.475 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.476 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.027 Collecting virtualenv
#12 4.091   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.174      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 75.4 MB/s eta 0:00:00
#12 4.217 Collecting distlib<1,>=0.3.7
#12 4.221   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.228      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 90.2 MB/s eta 0:00:00
#12 4.266 Collecting filelock<4,>=3.12.2
#12 4.269   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.306 Collecting platformdirs<5,>=3.9.1
#12 4.310   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.394 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.590 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.590 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.7s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      126912 kB
DirectMap2M:     8261632 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-05-12/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
##[endgroup]
[TIMING] core::build_steps::tool::ToolBuild { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 35.464
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/src/bootstrap/src/utils/shared_helpers.rs:46:
 /// Given an executable called `name`, return the filename for the
 /// executable for a particular target.
 pub fn exe(name: &str, target: &str) -> String {
-    if target.contains("windows")  || (cfg!(not(target_os = "cygwin")) && target.contains("cygwin"))
+    if target.contains("windows") || (cfg!(not(target_os = "cygwin")) && target.contains("cygwin"))
     {
         format!("{name}.exe")
     } else if target.contains("uefi") {
fmt: checked 6021 files
Build completed unsuccessfully in 0:01:16
  local time: Thu May 22 00:40:50 UTC 2025
  network time: Thu, 22 May 2025 00:40:50 GMT
##[error]Process completed with exit code 1.
```

</details>

onur-ozkan commented:
Thank you!

Could you please squash your commits into single commit? I am also not familiar with the Windows environment, so passing this to person that I kind a know that they are using Windows.

r? @jieyouxu

jeremyd2019 commented:
sure, that was the point of the `fixup!` commit.  I had a minor mishap pasting the change into github's web editor, and had an extra space in there.

jeremyd2019 committed:
make shared_helpers exe function work for both cygwin and non-cygwin hosts

On Cygwin, it needs to not append .exe, because /proc/self/exe (and
therefore std::env::current_exe) does not include the .exe extension,
breaking bootstrap's rustc wrapper.  On hosts other than Cygwin, it
*does* need to append .exe because the file really does have a .exe
extension, and non-Cygwin hosts won't be doing the same filename
rewriting that Cygwin does when looking for a file X but finding only
X.exe in its place.

mati865 reviewed (approved):
How about leaving a comment there? I feel like this condition will not be obvious, even to people familiar with Cygwin.

jieyouxu commented:
Yeah, please leave a comment. When I reviewed the original PR the change looked correct.

jieyouxu commented:
@rustbot author

jeremyd2019 commented:
@rustbot ready

jieyouxu reviewed (approved):
Thanks

jieyouxu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 1f862a82e2ed32a0c9a6dc33fb7cf32ed7195653 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 1f862a82e2ed32a0c9a6dc33fb7cf32ed7195653 -->
<!-- homu: {"type":"Approved","sha":"1f862a82e2ed32a0c9a6dc33fb7cf32ed7195653","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
